### ISSUE TITLE
Investigate possible Flutter engine cause of jitter

### ISSUE THREAD

**Issue by matthew-carroll** [Flutter-Bounty-Hunters/page_list_viewport#45]
We think that the cause of panning jitter might be due to behaviors within the Flutter engine. This ticket is to investigate possible causes and explore solutions.

The jitter seems to be more noticeable on higher refresh displays, such as the 90Hz Xiaomi Redmi tablet.

To recreate the jitter in a reproducible manner, use the floating action buttons at the lower right of the screen to animate panning up/down and left/right. You should see fairly frequent jitter, which looks a bit like rapid starting and stopping while panning.

**Comment by RossComputerGuy** [comment]
Which particular example Dart file should I run to investigate this?

**Comment by RossComputerGuy** [comment]
Alright, I tried `main_list` and noticed it's not affected but `main_minimal_viewport` is.

**Comment by RossComputerGuy** [comment]
I did some testing and this looks to be related to shaders. Following Flutter's documentation on [shader compilation jank](https://docs.flutter.dev/perf/shader) seems to have mitigated the problem mostly. I am not sure if that is exactly the issue but it's a lead. But even with using a precompiled shader, I notice in the Flutter DevTools that there's still some jank rendering.

![image](https://github.com/Flutter-Bounty-Hunters/page_list_viewport/assets/19699320/504ff76b-5f68-4f4a-87e0-f4f02de605fc)
![image](https://github.com/Flutter-Bounty-Hunters/page_list_viewport/assets/19699320/c70d5d0c-608d-4e35-bfab-0b2f08c55f01)

**Comment by matthew-carroll** [comment]
The jitter problem isn't visible in the profiler, in my experience. That occasional jank frame isn't good, but that's a different problem.

If you watch the UI as the panning animation runs, you'll see numerous jitter frames. And yet you won't see a bunch of back to back jank frames in the profiler. That's part of why this problem is difficult. Flutter doesn't seem to recognize that something isn't right.

**Comment by RossComputerGuy** [comment]
Interesting, so even with precompiled shaders it's jittery? I'll look more into it tomorrow by enabling the Skia tracing. Have you tried with running with Impeller? If not, I'll try tomorrow since I saw there seems to be others with this issue on Android and Impeller seems to be affected.

**Comment by matthew-carroll** [comment]
I haven't tried this experience with Impeller. We tried our direct PDF renderer with Impeller and it was way worse than Skia. The Android version of Impeller is still considered unstable and experimental. But you're welcome to give it a try.

You'll wanna keep a close eye on the screen as it pans. You might also want to download the Xodo app, which is a PDF renderer. You can pan around in there and feel the difference.

**Comment by RossComputerGuy** [comment]
Alright will do, I also came across [this page](https://docs.flutter.dev/perf/best-practices) which references a method called `saveLayer`. Might be worth looking into using that.

**Comment by matthew-carroll** [comment]
In the bitmap PDF renderer we're pushing texture layers, so there's nothing for us to save. That experience has the same jitter as this image based tiling experience.

**Comment by RossComputerGuy** [comment]
It appears the jitter is gone or reduced significantly under Impeller.

**Comment by matthew-carroll** [comment]
Is there a path forward where you could discover the relevant differences between Skia and Impeller, which lead to the elimination of the jitter? At least to the extent that we might point the Flutter team in that direction?

Eventually Impeller will be the norm, and we'll work under the expectation of using Impeller, but that time hasn't arrived yet for Android. So we're interested in a path that would result in fixes to Flutter's integration with Skia such that this problem goes away.

Do you think there's a path forward for you to help with that?

**Comment by RossComputerGuy** [comment]
The biggest change I've read for Impeller is that it compiles shaders before execution. I think if GC could not be executed during scrolling then it could lead to better performance.

**Comment by matthew-carroll** [comment]
Those sound like two very different problems. Is the problem repeated shader compilation? Or is the problem repeated GCs?

**Comment by RossComputerGuy** [comment]
I can't trace whether it is the shaders affecting it more. I did tracing and every time it starts doing a large paint, GC is running. GC is likely the large issue but shader recompilation can contribute somewhat.

**Comment by matthew-carroll** [comment]
Can you check the `PageListViewport` and see if we used the new API to disable GC? I know we used it somewhere, but I can't remember if we put it directly into `PageListViewport`, or into the bitmap PDF renderer.

If it's not already in `PageListViewport`, or in the controller, or something, then you could try using that new API to disable GC when panning begins, and then re-enable it a couple seconds after panning stops.

**Comment by RossComputerGuy** [comment]
Yeah, I don't know what the new GC API is.

**Comment by matthew-carroll** [comment]
From this post: https://medium.com/flutter/whats-new-in-flutter-3-7-38cbea71133c

Which links to this: https://github.com/dart-lang/sdk/commit/c6a1eb1b61844b2d733f9e2f4c7754f1920325d7

**Comment by RossComputerGuy** [comment]
Just checked and the new GC API is not being used at all. There isn't really any bindings to it from what I can see but adding the external references from [here](https://github.com/dart-lang/sdk/commit/c6a1eb1b61844b2d733f9e2f4c7754f1920325d7#diff-a04fd8c3f0edc9abfffbec083385d088bfb2aed22f1a8b32f469615502d95ee2R9710-R9724) should provide access to the right methods so it'll just be a matter of calling those.

**Comment by RossComputerGuy** [comment]
Just tried that and those methods aren't available, Flutter is likely using a custom Dart build with specific changes so it's likely they didn't expose the methods. I also tried with Impeller in release mode, it's smooth enough that I can barely tell there's any jitter. Unfortunately, it is hard to determine whether or not the actual FPS is an exact 90FPS consistently.

**Comment by matthew-carroll** [comment]
I'm looking for where we used that GC behavior. It's definitely there. It's just not easy to figure out.

WRT FPS - there's a dev option setting on the tablet where you can display the current frame rate on the screen, however that won't show the jitter, either.

The jitter is something that you have to see with your eyes. There's no other signal that we've found to confirm it's happening.

> it's smooth enough that I can barely tell there's any jitter

Ozzie said he tried with Impeller and the jitter is still there. When you run the vertical and horizontal animations, together, and it starts panning around the screen in an oval, you don't see recurring situations where you get a bunch of rapid, small visual stutters?

**Comment by RossComputerGuy** [comment]
> I'm looking for where we used that GC behavior. It's definitely there. It's just not easy to figure out.

Gotcha, I'm not too familiar with GC so I wasn't sure.

> Ozzie said he tried with Impeller and the jitter is still there.

Yeah if I stare long enough at the screen, I can see some jitter but it's not a ton. It seems to be at certain frames. Debug and profile have the most jitter with Impeller, release has it more difficult for me to notice.

**Comment by matthew-carroll** [comment]
If it's easier to reproduce and recognize with Skia, then let's go back to that. Impeller isn't really relevant to us, unless Impeller has no jitter, and you're able to learn something from Impeller.

But we're building against the Skia version. That's where we're suffering the issue, and that's what customers will be stuck with. So we're trying to achieve one of the following:

1. Ideally, find and fix the engine/skia interaction that's causing the jitter, or
2. Accumulate sufficient contextual information so that we can point the Flutter team to a narrow problem definition in the hope that they can fix it

Do you feel that you have possible paths to reach those goals, do you feel like it's completely unknown what to do, or where to look?

**Comment by RossComputerGuy** [comment]
Alright, I'll see what I can do to figure out how GC is affecting rendering through Skia. If it isn't GC then I'm not sure what else it could be since I haven't been able to get useful information from enabling Skia tracing.

**Comment by matthew-carroll** [comment]
I don't know the structure of the engine at that level, but another thing you might try to look into is when/where/how textures and images are positioned for final rasterization. There's a chance that they might be placed with some kind of rounding error, which could be perceived as jitter. A low chance, but a chance that's what's going on.

WRT GCs, if there are a bunch of GCs happening, it's probably worth understanding why they're happening. If we're panning around in a circle, we're not creating a bunch of new widgets, and we're not loading new images. What's getting recycled?

**Comment by RossComputerGuy** [comment]
Yeah, that's ultimately what I'm trying to understand with the GC. I see the memory usage rise from 69MB to 80MB and repeats for every frame. I'll look and see if there's some sort of way of doing verbose GC so I can figure where things are allocating and where things are being disposed of.

**Comment by matthew-carroll** [comment]
Here's where we utilized the GC controls: https://github.com/Flutter-Bounty-Hunters/page_list_viewport/pull/14/files

**Comment by matthew-carroll** [comment]
Also, when it comes to root causing, even if you do something like hack Dart or Flutter, itself, by raising the max memory threshold way up high, that's fine too. We need proof about where this problem is coming from....

**Comment by RossComputerGuy** [comment]
Alright, I'll use that. I decided to add a break [here](https://github.com/Flutter-Bounty-Hunters/page_list_viewport/blob/main/example/lib/pages_with_tiled_images/main_page_list_images_inspector.dart#L178) and discovered this is called right before scrolling to the next page. Is the `PageListViewport` class not caching the entries?

**Comment by matthew-carroll** [comment]
There's quite a bit of complexity here, so you'll have to be more specific. What, precisely, would you expect to be cached?

**Comment by RossComputerGuy** [comment]
Is `PageWithTileImages` not being cached in the `PageListViewport`? It seems like its disposing and creating a new one every time it scrolls to a different page.

**Comment by matthew-carroll** [comment]
You're asking if the `Widget` instance is being cached within the page list viewport's `Element`?

**Comment by RossComputerGuy** [comment]
Yeah, that's essentially what I am asking. The `PageWithTileImages` is being disposed and recreated every scroll. I replaced the builder method with this and the average memory usage decreased by about 5MB.
```dart
if (pages.elementAtOrNull(pageIndex) == null) {
   pages.insert(pageIndex, PageWithTileImages(
     tileCache: _tileCache,
     pageIndex: pageIndex,
     naturalSize: _naturalPageSize,
     levelZeroTileFilterQuality: FilterQuality.medium,
     showTileBounds: true,
   ));
}
return pages[pageIndex];
```

**Comment by matthew-carroll** [comment]
Some amount of rebuilding is there because the page builder is given the current viewport bounds. But it wouldn't surprise me if there are still various places where that can be optimized. Feel free to hack any changes you want, if it helps you root cause the issue. Just make sure that if you end up solving the issue, you're able to track that solution back to a specific change that you made.

**Comment by RossComputerGuy** [comment]
Found this in a jank frame reported by the performance measuring tool. I think the viewport optimizer timer might need to be increased, I'll mess around with that and see if it does anything.
![image](https://github.com/Flutter-Bounty-Hunters/page_list_viewport/assets/19699320/db9e4096-0003-4d36-b7dd-ac13e8cf2004)

**Comment by RossComputerGuy** [comment]
I changed the cancel timer in the performance optimizer from 3 to 45 and noticed a more fluid movement. I chose 45 because that is the delay of the longest frame event listed in the DevTools. Currently the tools in Android Studio report an average of 68 - 70 FPS. The web tools reports 90 FPS consistent average. I also notice GC is running a lot less. The jitter does not seem to appear very rarely in debug mode. Release mode has fully no jitter. I think this shows that GC is definitely running too often and that should be adjusted wherever that comes from in Flutter.

**Comment by matthew-carroll** [comment]
> I changed the cancel timer in the performance optimizer from 3 to 45

Are you saying 3s to 45s? We certainly can't turn off GC for 45 seconds, but why do you think such a large number would have an impact?

> noticed a more fluid movement

Does this mean the jitter is gone?

> Android Studio report an average of 68 - 70 FPS. The web tools reports 90 FPS consistent average

It's really important that you understand that the metrics these tools are reporting don't reflect this jitter problem. I've mentioned that a few times. So going back to those numbers as a signal is likely to lead you down bad paths. We can't see the problem in those numbers to begin with, so we're not going to see the solution in those numbers, either.

As for the difference in those two numbers, I'm not sure why. If you see a way to reproduce that, you might want to file an issue with Flutter.

> The jitter does not seem to appear much in debug mode. Release mode has fully no jitter.

Are you talking about after this change?

If you think changing the value has largely solved the problem, please put up a PR with that change so that others can run it and see if it's resolved.

**Comment by RossComputerGuy** [comment]
> Are you saying 3s to 45s? We certainly can't turn off GC for 45 seconds, but why do you think such a large number would have an impact?

I noticed that the impact of GC being off for longer didn't lead to any significant impact because when the timer executes, it's only with the spike which could be related to something with shaders.

> Does this mean the jitter is gone?

Yes, I cannot see any jitter in release mode, debug mode has a very small amount of jitter.

> It's really important that you understand that the metrics these tools are reporting don't reflect this jitter problem.

Yeah, I've been using those metrics to help get some idea but not determine the true performance. I've been monitoring the frame rate listed along with paying very close attention to the display.

> Are you talking about after this change?

Yes, changing the timer time from 3 seconds to 45 seconds causes the GC to not reset as quickly. It's just debug mode is slightly slower and so there is some jitter because it is not a release build so it isn't optimized as much at compile time.

> If you think changing the value has largely solved the problem, please put up a PR with that change so that others can run it and see if it's resolved.

Will do.

**Comment by matthew-carroll** [comment]
@RossComputerGuy I still didn't follow the first part about timing. You used the numbers "3" and "45", can you clarify that you're talking about the number of seconds after final interaction?

I also didn't understand this response:

> I noticed that the impact of GC being off for longer didn't lead to any significant impact because when the timer executes, it's only with the spike which could be related to something with shaders.

**Comment by RossComputerGuy** [comment]
So basically with the timer [here](https://github.com/Flutter-Bounty-Hunters/page_list_viewport/blob/main/lib/src/page_list_performance_optimizer.dart#L80), it is basically a timeout timer. Majority of the frames Flutter is rendering occurs in less than 3 seconds so we never execute the timer. However, with the frames that have a huge spike in time, those cannot be fixed because it appears it has to do with shaders or initial framing. Since those frames take much longer, the timer executes while we're still rendering. This leads to Flutter doing more work too soon and so changing this timeout from 3 to 45 fixes this issue. 45 was chosen because the Flutter frame analysis tool reports the longer frames take roughly 45 seconds. With the events Flutter was dispatching, GC is now running after the frames are rendered instead of in the middle of frame rendering.

**Comment by matthew-carroll** [comment]
I still don't understand this explanation. The timer goes off after 3 seconds of *inactivity*. There shouldn't be any frames rendering when that timer goes off. Three seconds is already way more time than Flutter needs to finish any pending work. Flutter can only ever be, at most, 1 frame behind. The reason we chose 3 seconds was so that if the user pauses momentarily, and then starts panning again, we won't trigger a GC in between.

So I still don't understand what you mean about frames still rendering, and Flutter still doing work after 3 seconds of inactivity, and I still don't understand why making that number larger would have any impact.

**Comment by RossComputerGuy** [comment]
I'm not sure then either but it's working and I don't see GC events as frequently in the event log. The memory usage graph also has less spikes. It used to be GC was running more than every second, now it's almost every 2 seconds.

**Comment by matthew-carroll** [comment]
Ok. Please put up a PR so I can take a look.

**Comment by RossComputerGuy** [comment]
Alright, I made a PR.

**Comment by matthew-carroll** [comment]
@RossComputerGuy I ran your PR with the "Image tiles" entrypoint. I ran it in release mode and profile mode. When I run the oval panning animation, I see just as much jitter as before. I didn't notice any improvements.

**Comment by matthew-carroll** [comment]
BTW, when you run the horizontal and vertical animations, together, causing the most motion and jitter, does your Android Studio frame profiler show nearly constant jank frames? If so, does your Dev Tools also show that many jank frames? I'm getting different results.

<img width="340" alt="Screen Shot 2023-05-30 at 3 42 52 PM" src="https://github.com/Flutter-Bounty-Hunters/page_list_viewport/assets/7259036/f4376292-8126-4c42-a238-5c45866deee7">
<img width="1446" alt="Screen Shot 2023-05-30 at 3 43 01 PM" src="https://github.com/Flutter-Bounty-Hunters/page_list_viewport/assets/7259036/24f44b41-5a9e-4fb1-982c-c937fe74ed3d">

**Comment by RossComputerGuy** [comment]
Yeah so Android Studio shows completely different frame rendering times than the Flutter DevTools for me as well. I am not sure why though. I am also not sure why you're getting jitter but I don't seem to be. I recorded the screen with `adb shell screenrecord --bugreport` and it says a few hundred FPS, granted I don't know the metrics used for that frame counter but if there is jitter then it should show up in the video. [It's available here](https://youtu.be/3xWm0L3Y3K8). I was able to see the jitter in progress even with setting the 5s delay option in the devtools. I'll look into compiling the Flutter Engine and trying to get a Skia trace.

**Comment by matthew-carroll** [comment]
In general we haven't been able to demonstrate the issue with video because frame rates are different everywhere along the path to generating videos.

However, ironically, I can see jitter in your example video. There are very slight stutters, especially when the that panning motion slows down at the end, and speeds back up in the reverse direction.

You might also see the jitter more clearly if you zoom in to get a variety of image tiles on the screen and then run the animation. Also, consider playing both animations together, to get an oval panning motion, which might reveal even more.

**Comment by RossComputerGuy** [comment]
Hmm alright, I'll try analyzing frames and compare it with Xodo. Slowing the video down to 1/4 speed, I can see that slight jitter you're talking about now. If you record a video using the same command on your device, is the jitter the same or worse than it is for me?

**Comment by RossComputerGuy** [comment]
Looking into this more, seems like raster is causing the issues. I couldn't get the integrated Perfetto UI to work but I was able to capture a trace.
![image](https://github.com/Flutter-Bounty-Hunters/page_list_viewport/assets/19699320/85e45f40-f475-47da-a5aa-c540dc9b079b)
The large `drawFrame` event is 10ms, could that be a lead onto this issue?

**Comment by matthew-carroll** [comment]
Your guess is as good as mine. My first question would be how often does this situation happen in the trace? We know that we're seeing jitter that happens over and over, for many frames. Do you see an equivalent number of these trace results?

If you do, then the question goes to the root cause. You've identified a long `draw` call, but it's not clear to me what's actually causing that `draw` call to take so long. I think we'd need to figure out why.

**Comment by dnfield** [comment]
So as a disclaimer, I'm not able to super reliably reproduce the jittery issue and when I can reproduce it I'm not necessarily the best at visually detecting it.

That said, I tried the following patch locally and it seems to help:

```diff
diff --git a/shell/platform/android/platform_view_android.cc b/shell/platform/android/platform_view_android.cc
index b89c23a368..2ff8fc63ae 100644
--- a/shell/platform/android/platform_view_android.cc
+++ b/shell/platform/android/platform_view_android.cc
@@ -204,6 +204,12 @@ void PlatformViewAndroid::NotifyChanged(const SkISize& size) {
   latch.Wait();
 }
 
+PointerDataDispatcherMaker PlatformViewAndroid::GetDispatcherMaker() {
+  return [](DefaultPointerDataDispatcher::Delegate& delegate) {
+    return std::make_unique<SmoothPointerDataDispatcher>(delegate);
+  };
+}
+
 void PlatformViewAndroid::DispatchPlatformMessage(JNIEnv* env,
                                                   std::string name,
                                                   jobject java_message_data,
diff --git a/shell/platform/android/platform_view_android.h b/shell/platform/android/platform_view_android.h
index 300f3ab9b6..59d3618990 100644
--- a/shell/platform/android/platform_view_android.h
+++ b/shell/platform/android/platform_view_android.h
@@ -117,6 +117,9 @@ class PlatformViewAndroid final : public PlatformView {
     return platform_message_handler_;
   }
 
+  // |PlatformView|
+  PointerDataDispatcherMaker GetDispatcherMaker() override;
+
  private:
   const std::shared_ptr<PlatformViewAndroidJNI> jni_facade_;
   std::shared_ptr<AndroidContext> android_context_;
```

@matthew-carroll are you able to apply that to an engine checkout and see if it fixes things for you? If so I could look into making it landable upstream.

**Comment by dnfield** [comment]
(that patch is meant to help with irregularities when the refresh rate on the screen and the touch input device are different and may get out of sync)

**Comment by matthew-carroll** [comment]
@dnfield we can try the patch, but I'm not sure how it's relevant?

We are currently reproducing the jitter with animations that are tied to a couple of buttons in the demo. There aren't any touch input events happening...

**Comment by dnfield** [comment]
Ah, sorry I had missed the buttons. Touch input could be something else on the device I'm looking at.

**Comment by dnfield** [comment]
For the buttons one, disabling the raster cache seems to help. Checking into that a bit more.

**Comment by dnfield** [comment]
Also, your trace is showing `eglSwapBuffersWithDamageKHR` above - but that call was removed in a recent version of Flutter (https://github.com/flutter/engine/pull/40898) and the commit containing it made it into the 3.10 stable.

**Comment by matthew-carroll** [comment]
@dnfield I'm definitely on 3.10 stable, but @RossComputerGuy can you please confirm that you're on the latest Flutter stable version? If you're not, can you upgrade and then rerun your traces?

**Comment by RossComputerGuy** [comment]
I'm on Flutter master, should I be using a tagged version like 3.10?

**Comment by matthew-carroll** [comment]
master should be fine

**Comment by dnfield** [comment]
Ok, it looks like Android internally is calling `eglSwapBuffersWithDamageKHR` from `eglSwapBuffers` ... on some devices and versions of Android at least.

Right now I think we're sometimes running into a problem where we're double stuffing the buffer and failing to catch up. Each frame gets delayed. This is a known issue that's not got a really good solution unfortunately.

**Comment by matthew-carroll** [comment]
>  This is a known issue that's not got a really good solution unfortunately.

Can you elaborate on that? Does this mean there's nothing we can do on our end to fix it? Does this mean that no one is going to address this on the Flutter side?

Our client's goal has always been to deliver a premium reading experience. This bug makes that impossible. A reader that jitters every time the user touches the screen will never be perceived as premium. As a result, this seemingly minor issue is catastrophic for the client. So it's a very big deal if this problem is essentially a shoulder shrug from Flutter.

**Comment by dnfield** [comment]
I mean the thing I think I'm seeing is a known problem that no one has quite complained about loudly enough yet.

There were two or three attempts at landing something that shoudl fix it, but none of them worked and they ended up causing other, bigger problems instead.

We were able to fix this on iOS by using transactions. But I don't know that the same are available on Android (particularly in the GLES backend). I could try to ask around but it's not the first time I've tried to figure out what's going on with this particular problem.

It's possible I'm missing something, that there's one or more other issues contributing to this (I do think the touch input is a problem on at least some android devices at this point for example).

**Comment by dnfield** [comment]
I can't seem to reproduce the jitter on iOS, which further seems to support my hypothesis.

**Comment by dnfield** [comment]
On a Samsung S6 tablet, I can see this example frequently exceeding budget on the raster thread.

If I switch rendering to a SurfaceView instead of a TextureView, things get much better. There's still some occasional jitteriness but over all it looks much better in tracing and visually to me (specifically, _don't do this_: https://github.com/Flutter-Bounty-Hunters/page_list_viewport/blob/cd74fce4a3d02e33dbdac0adafa5376b0fe8c9b0/example/android/app/src/main/kotlin/com/example/example/MainActivity.kt#L7).

**Comment by dnfield** [comment]
(Basically, you should _never_ use RenderMode.texture unless you absolutely must)

**Comment by dnfield** [comment]
Second big problem:

The images you're drawing in the tiles are _huge_. As in, every single image your drawing is larger than the physical size of the screen. The framework has debug checking for this, but you're bypassing that by directly using `canvas.drawImage`.

This is causing tons of GPU work that is completely unnecessary, not to mention lots of memory usage.

**Comment by dnfield** [comment]
You can help address that by using a `ResizeImage` wrapped around your `AssetImage` in image_cache.dart.

**Comment by matthew-carroll** [comment]
@RossComputerGuy can you try both of Dan's recommendations? Switch back to `SurfaceView` in the Android `Activity` and also resize the images?

**Comment by RossComputerGuy** [comment]
Yeah, I'll try those out. If its worth it, I want to get a trace dump in Perfetto.

**Comment by RossComputerGuy** [comment]
I just tried this out, I switched to surface rendering and removed the `* 72` part in the `_naturalPageSize` variable. Performance is now excellent. I don't see any jitter and the screen recording shows there's no jitter. I decided to check the framerate in different things and this is what I got for average FPS's:

- Screen recording: 1k
- Android Studio: 88
- Flutter DevTools: 89

This was in debug mode and I tried it with a prebuilt engine on tag `3.10.3` since I am on my laptop for the next month and I don't want to recompile the engine if needed.

**Comment by matthew-carroll** [comment]
Can you post a PR with those changes so that @kirkbyo and I can run the demo and see what happens?

**Comment by RossComputerGuy** [comment]
Pushed a PR

**Comment by matthew-carroll** [comment]
I tried the PR that goes back to `SurfaceView` and removes the `72` multiplier.

First, the experience is broken when the `72` is removed. I can't pinch to zoom at all. This likely indicates that the reported findings were based on the lowest-load experience, which is one tile per page.

I added `72` back, zoomed in to get a bunch of tiles, ran the automatic animations, and I'm still getting as much jitter as ever.

**Comment by dnfield** [comment]
I'm not sure what the 72 is referring to, but the real kicker is to make sure that you're not decoding your images at full resolution - try to decode them as close as possible to the resolution you'll actually display them at. A little bigger is ok, but right now they're getting decoded as something like 5k resolution to be displayed on a 1080 screen.

**Comment by matthew-carroll** [comment]
@dnfield - Makes sense.

@RossComputerGuy can you update the PR so that this is accomplished? We still need full pinch to zoom capability, and we want to test the jitter with as many tiles on screen as possible.

**Comment by RossComputerGuy** [comment]
I've updated the PR, I made it a 5th of what it originally was and that works.
